lib/services/local/headlessauthn_watcher_test.go

```diff
@@ -18,283 +18,281 @@ package local_test
-			return ha.User != "", nil
-		})
+		sub, err := s.watcher.Subscribe(ctx, pubUUID)
+		require.NoError(t, err)
+		t.Cleanup(sub.Close)
+
+		ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+		t.Cleanup(cancel)
+
+		headlessAuthnCh := make(chan *types.HeadlessAuthentication, 1)
+		errC := make(chan error, 1)
+		go func() {
+			ha, err := s.watcher.WaitForUpdate(ctx, sub, func(ha *types.HeadlessAuthentication) (bool, error) {
+				return ha.User != "", nil
+			})
+			headlessAuthnCh <- ha
+			errC <- err
+		}()
 
-		// Emit put event that passes the condition.
-		stub, err := types.NewHeadlessAuthenticationStub(pubUUID, time.Now())
+		// Make an update that passes the condition.
+		stub, err := s.identity.CreateHeadlessAuthenticationStub(ctx, pubUUID)
 		require.NoError(t, err)
-		stub.User = "user"
 
-		waitForPutEvent(t, s, stub, firstEventReceivedC)
+		replace := *stub
+		replace.User = "user"
+		replace.PublicKey = []byte(sshPubKey)
+		s.identity.CompareAndSwapHeadlessAuthentication(ctx, stub, &replace)
 
 		require.NoError(t, <-errC)
-		require.Equal(t, stub, <-headlessAuthnCh)
+		require.Equal(t, &replace, <-headlessAuthnCh)
 	})
 
-	t.Run("WaitEventWithConditionUnmet", func(t *testing.T) {
+	t.Run("ConditionUnmet", func(t *testing.T) {
 		t.Parallel()
-		s := newTestEnv(t)
+		s := newHeadlessAuthenticationWatcherTestEnv(t)
 
-		waitCtx, waitCancel := context.WithCancel(ctx)
-		t.Cleanup(waitCancel)
+		sub, err := s.watcher.Subscribe(ctx, pubUUID)
+		require.NoError(t, err)
+		t.Cleanup(sub.Close)
 
-		headlessAuthnCh, firstEventReceivedC, errC := waitInGoroutine(waitCtx, t, s.watcher, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) {
-			return ha.User != "", nil
-		})
+		ctx, cancel := context.WithCancel(ctx)
+		t.Cleanup(cancel)
 
-		// Emit put event that doesn't pass the condition (user not set). The waiter should ignore these events.
-		stub, err := types.NewHeadlessAuthenticationStub(pubUUID, time.Now())
-		require.NoError(t, err)
+		headlessAuthnCh := make(chan *types.HeadlessAuthentication, 1)
+		errC := make(chan error, 1)
+		go func() {
+			ha, err := s.watcher.WaitForUpdate(ctx, sub, func(ha *types.HeadlessAuthentication) (bool, error) {
+				return ha.User != "", nil
+			})
+			errC <- err
+			headlessAuthnCh <- ha
+		}()
 
-		waitForPutEvent(t, s, stub, firstEventReceivedC)
+		// Make an update that doesn't pass the condition (user not set). The waiter should ignore these events.
+		_, err = s.identity.CreateHeadlessAuthenticationStub(ctx, pubUUID)
+		require.NoError(t, err)
 
 		// Ensure that the waiter did not finish with the condition unmet.
 		select {
 		case err := <-errC:
-			t.Errorf("Expected waiter to continue but instead the waiter returned with err: %v", err)
+			t.Fatalf("Expected waiter to continue but instead the waiter returned with err: %v", err)
 		default:
-			waitCancel()
+			cancel()
```

Comment:
Won't this almost instantly fall into the cancel? Should we give some time for the waiter to (not) process the update?